File: attention/lower_triangular_inverse/pta/lower_triangular_inverse/converter/lower_triangular_inverse.py
# Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import (
    Any, Callable, ContextManager, Iterator, List, Literal, NamedTuple, Optional, Sequence, Tuple, TypeVar,
    Union, overload,
)
import torch
import torch_npu
import torchair
from torch import Generator, contiguous_format, inf, strided, SymInt
from torch.types import Device, Number, _bool, _complex, _device, _dtype, _float, _int, _layout, _qscheme, _size
from torchair._ge_concrete_graph import ge_apis as ge
from torchair._ge_concrete_graph.fx2ge_converter import declare_supported, register_fx_node_ge_converter
from torchair.ge._ge_graph import Tensor, TensorSpec, DataType, TensorType
from torchair._ge_concrete_graph.supported_declaration import _TypedTensor, F32, F16, F64, I32, I16, I64, I8, U8, \
    BOOL, Support
from torchair._ge_concrete_graph.utils import dtype_promote
from torchair.ge import attr
from torch.library import Library, impl
from typing import Any, List, Tuple, Union, Callable, Optional
from torchair.ge._ge_graph import get_default_ge_graph, next_unique_name, auto_convert_to_tensor, compat_as_bytes, \
    get_invalid_desc

@register_fx_node_ge_converter(torch.ops.custom.npu_lower_triangular_inverse.default)
def conveter_npu_lower_triangular_inverse(
    x: Tensor,
    meta_outputs: TensorSpec = None
):

    return torchair.ge.custom_op(
        "LowerTriangularInverse",
        inputs={
            "x": x
        },
        outputs={
            'out'
        }
    )

File: attention/lower_triangular_inverse/pta/lower_triangular_inverse/csrc/lower_triangular_inverse.cpp
/* *
 * Copyright (c) Huawei Technologies Co., Ltd. 2025. All rights reserved.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <iostream>
#include <torch/library.h>
#include "../../../../csrc_base/ops_common.h"

namespace custom {
constexpr int X_DIM_NUM = 5;

using npu_preparation = at_npu::native::OpPreparation;

at::Tensor npu_lower_triangular_inverse(const at::Tensor& x) {
    TORCH_CHECK(x.dim() == X_DIM_NUM, "value dim should be 5, but actual is ", x.dim());
    at::Tensor result = at::empty_like(x);

    EXEC_NPU_CMD_V1(aclnnLowerTriangularInverse, x, result);
    return result;
}

at::Tensor npu_lower_triangular_inverse_meta(const at::Tensor& x) {
    TORCH_CHECK(x.dim() == X_DIM_NUM, "value dim should be 5, but actual is ", x.dim());
    at::Tensor result = at::empty_like(x);

    return result;
}
}

// 为NPU设备注册前向实现
TORCH_LIBRARY_IMPL(custom, PrivateUse1, m) {
    m.impl("npu_lower_triangular_inverse", &custom::npu_lower_triangular_inverse);
}

TORCH_LIBRARY_IMPL(custom, Meta, m) {
    m.impl("npu_lower_triangular_inverse", &custom::npu_lower_triangular_inverse_meta);
}